<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1-Введение" data-toc-modified-id="1-Введение-1">1 Введение</a></span><ul class="toc-item"><li><span><a href="#1.1-Клиент" data-toc-modified-id="1.1-Клиент-1.1">1.1 Клиент</a></span></li><li><span><a href="#1.2-Цель" data-toc-modified-id="1.2-Цель-1.2">1.2 Цель</a></span></li><li><span><a href="#1.3-Планируемые-этапы-исследования" data-toc-modified-id="1.3-Планируемые-этапы-исследования-1.3">1.3 Планируемые этапы исследования</a></span></li></ul></li><li><span><a href="#2-Подготовка-данных" data-toc-modified-id="2-Подготовка-данных-2">2 Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#2.1-Описание-данных" data-toc-modified-id="2.1-Описание-данных-2.1">2.1 Описание данных</a></span></li><li><span><a href="#2.2-Подключение-к-базе-данных" data-toc-modified-id="2.2-Подключение-к-базе-данных-2.2">2.2 Подключение к базе данных</a></span></li></ul></li><li><span><a href="#3-Исследование-таблиц" data-toc-modified-id="3-Исследование-таблиц-3">3 Исследование таблиц</a></span><ul class="toc-item"><li><span><a href="#3.1-Количество-вышедших-книг-после-1-января-2000-года" data-toc-modified-id="3.1-Количество-вышедших-книг-после-1-января-2000-года-3.1">3.1 Количество вышедших книг после 1 января 2000 года</a></span></li><li><span><a href="#3.2-Количество-обзоров-и-средняя-оценка-для-каждой-книги" data-toc-modified-id="3.2-Количество-обзоров-и-средняя-оценка-для-каждой-книги-3.2">3.2 Количество обзоров и средняя оценка для каждой книги</a></span></li><li><span><a href="#3.3-Издательство,-выпустившее-наибольшее-число-книг-толще-50-страниц(исключаются-брошюры)" data-toc-modified-id="3.3-Издательство,-выпустившее-наибольшее-число-книг-толще-50-страниц(исключаются-брошюры)-3.3">3.3 Издательство, выпустившее наибольшее число книг толще 50 страниц(исключаются брошюры)</a></span></li><li><span><a href="#3.4-Автор-с-самой-высокой-средней-оценкой-книг(c-50-и-более-оценками)" data-toc-modified-id="3.4-Автор-с-самой-высокой-средней-оценкой-книг(c-50-и-более-оценками)-3.4">3.4 Автор с самой высокой средней оценкой книг(c 50 и более оценками)</a></span></li><li><span><a href="#3.5-Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="3.5-Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5">3.5 Среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#4-Выводы" data-toc-modified-id="4-Выводы-4">4 Выводы</a></span></li></ul></div>

# Проект: Анализ данных сервиса для чтения книг(SQL)

## 1 Введение

### 1.1 Клиент

Сервис для чтения книг по подписке

### 1.2 Цель

Выполнить анализ данных, согласно заданию:
- Посчитайте, сколько книг вышло после 1 января 2000 года
- Для каждой книги посчитайте количество обзоров и среднюю оценку
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

### 1.3 Планируемые этапы исследования

- Подключение к БД
- Анализ схемы БД и таблиц
- Написание запросов по поставленным задачам
- Написание выводов



## 2 Подготовка данных

### 2.1 Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

![structure](https://i.ibb.co/PhKF9fJ/Untitled-33.png)

### 2.2 Подключение к базе данных

In [1]:
# импортируем библиотеки

import pandas as pd 
from sqlalchemy import create_engine 


# устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                         db_config['pwd'], 
                                                         db_config['host'], 
                                                         db_config['port'], 
                                                         db_config['db'])

# сохраняем коннектор

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [21]:
# Информация о типе данных

query = '''

SELECT
  COLUMN_NAME, DATA_TYPE 
FROM
  INFORMATION_SCHEMA.COLUMNS 
WHERE
  TABLE_NAME = 'books'

;

'''

pd.io.sql.read_sql(query, con = engine)

column_name data_type
0           book_id   integer
1         author_id   integer
2             title      text
3         num_pages   integer
4  publication_date      date
5      publisher_id   integer

## 3 Исследование таблиц

### 3.1 Количество вышедших книг после 1 января 2000 года

In [2]:
query = '''

select count(book_id)

from books 
WHERE EXTRACT(YEAR FROM CAST(publication_date AS DATE)) >= 2000
;

'''

pd.io.sql.read_sql(query, con = engine)

count
0    821

- Количество книг вышедших после 1 января 2000: 821

### 3.2 Количество обзоров и средняя оценка для каждой книги

In [3]:
query = '''

select b.title,
        avg(rt.rating) avg_rating,
        count(distinct(rw.text)) reviews_count

from books as b
left join ratings as rt on rt.book_id = b.book_id
left join reviews as rw on rw.book_id = b.book_id
group by b.book_id
order by avg(rt.rating) DESC
-- order by count(distinct(rw.text)) DESC

;

'''

pd.io.sql.read_sql(query, con = engine)

title  avg_rating  \
0        Arrows of the Queen (Heralds of Valdemar  #1)        5.00   
1    The Walking Dead  Book One (The Walking Dead #...        5.00   
2                                      Light in August        5.00   
3    Wherever You Go  There You Are: Mindfulness Me...        5.00   
4    Captivating: Unveiling the Mystery of a Woman'...        5.00   
..                                                 ...         ...   
995  The World Is Flat: A Brief History of the Twen...        2.25   
996                  His Excellency: George Washington        2.00   
997                                      Drowning Ruth        2.00   
998                                              Junky        2.00   
999                               Harvesting the Heart        1.50   

     reviews_count  
0                2  
1                2  
2                2  
3                2  
4                2  
..             ...  
995              3  
996              2  
997              3  
998              2  
999              2  

[1000 rows x 3 columns]

In [4]:
query = '''

select avg(avg_rating) avg_rating_all,
        avg(reviews_count) avg_reviews_all

from
(
select b.title,
        avg(rt.rating) avg_rating,
        count(distinct(rw.text)) reviews_count

from books as b
left join ratings as rt on rt.book_id = b.book_id
left join reviews as rw on rw.book_id = b.book_id
group by b.book_id

) a1

;

'''

pd.io.sql.read_sql(query, con = engine)

avg_rating_all  avg_reviews_all
0        3.898973            2.793

Количество отзывов:
- Максимально: 7
- Минимально: 0
- Среднее количество отзывов на книгу: 2.8


Рейтинг:
- Максимальный: 5
- Минимальный: 1.5
- Средняя оценка книг: 3.9

### 3.3 Издательство, выпустившее наибольшее число книг толще 50 страниц(исключаются брошюры)

In [5]:
query = '''

select publisher,
        count(book_id)
        
from books as b
left join publishers as p on p.publisher_id = b.publisher_id
where num_pages >50
group by publisher
order by count(book_id) DESC
limit 1

'''

pd.io.sql.read_sql(query, con = engine)

publisher  count
0  Penguin Books     42

- Издательство выпустившее максимальное количество книг(от 50 страниц): Penguin Books, 42

### 3.4 Автор с самой высокой средней оценкой книг(c 50 и более оценками)

In [6]:
query = '''

with
a1 as
(
select at.author,
        avg(rt.rating),
        count(rt.rating)

from books as b
left join ratings as rt on rt.book_id = b.book_id
left join authors as at on at.author_id = b.author_id
group by at.author, b.book_id
having count(rt.rating) >= 50
order by avg(rt.rating) DESC
)

select author,
        avg(avg)
        
from a1
group by author

;

'''

pd.io.sql.read_sql(query, con = engine)

author       avg
0                          J.K. Rowling/Mary GrandPré  4.283844
1                                           Dan Brown  3.754540
2                                       J.D. Salinger  3.825581
3                                     William Golding  3.901408
4                                   Louisa May Alcott  4.192308
5            Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
6                                     Stephenie Meyer  3.662500
7   William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
8                                        Rick Riordan  4.080645
9             George Orwell/Boris Grabnar/Peter Škerl  3.729730
10                                     John Steinbeck  3.622951
11                                     J.R.R. Tolkien  4.258446
12                                         Lois Lowry  3.750000
13                  Markus Zusak/Cao Xuân Việt Khương  4.264151

- Автор с самой высокой средней оценкой книг(c 50 и более оценками): J.K. Rowling/Mary GrandPré	, 4.28

### 3.5 Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [7]:
query = '''

with
a2 as
(
with
a1 as
(
select username
--        count(rating)

from ratings 
group by username
having count(rating) >= 50
order by count(rating) desc
) 

select rw.username,
       count(text)

from reviews as rw
join a1 on rw.username = a1.username

group by rw.username
order by rw.username
)

select avg(count)
from a2


;

'''

pd.io.sql.read_sql(query, con = engine)

avg
0  24.222222

- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24.22

## 4 Выводы

- Количество книг вышедших после 1 января 2000: 821


- Издательство выпустившее максимальное количество книг(от 50 страниц): Penguin Books, 42


- Автор с самой высокой средней оценкой книг(c 50 и более оценками): J.K. Rowling/Mary GrandPré	, 4.28


- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24.22


Количество отзывов:
- Максимально: 7
- Минимально: 0
- Среднее количество отзывов на книгу: 2.8


Рейтинг:
- Максимальный: 5
- Минимальный: 1.5
- Средняя оценка книг: 3.9